ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs

```diff
@@ -112,16 +120,28 @@ jobPromoteColdPeer PeerSelectionActions{peerStateActions = PeerStateActions {est
   where
     handler :: SomeException -> Completion m peeraddr peerconn
     handler e =
-      Completion $ \st _now -> Decision {
-        decisionTrace = TracePromoteColdFailed peeraddr e,
-        decisionState = st {
-                          knownPeers            = KnownPeers.incrementFailCount
-                                                    peeraddr (knownPeers st),
-                          inProgressPromoteCold = Set.delete peeraddr
-                                                    (inProgressPromoteCold st)
-                        },
-        decisionJobs  = []
-      }
+      Completion $ \st now ->
+        let (failCount, knownPeers') = KnownPeers.incrementFailCount
+                                         peeraddr
+                                         (knownPeers st)
+
+            -- exponential backoff: 5s, 10s, 20s, 40s, 80s, 160s.
+            delay :: DiffTime
+            delay = fromIntegral $
+              2 ^ (pred failCount `min` maxColdPeerRetryBackoff) * baseColdPeerRetryDiffTime
```

Comment:
> Could we structure this so it's more similar to the existing backoff code in the RootPeers job handler? It'd make it easier to see that we're doing the same thing, or what the differences the the parameters are.

You'd like to count number of consecutive failures or successes? This could be a useful metric.

> Why do we want to pick 2^5 * 5 = 160 as the max time limit? For the root peers backoff we backed off for a max of around an hour. I'm not saying this is wrong, just that we should probably say something about why we chose these numbers.

Let's see: if we have 100 cold peers and 30% is offline, then will will try to reconnect to one of them every: 160/30=5.33s (assuming that the policy keeps asking for the 30% offline peers and assuming that the governor is constantly looking for more warm peers).   Do you think we should have a slower asymptotic behaviour?

Reply:
I really just meant the structure of the code should be similar, since as far as I understand it they're using the same way of counting failure and same backoff strategy.

Lets include that explanation in the code next to where we pick the backoff parameters.

Reply:
I pushed refactoring in this [commit](https://github.com/input-output-hk/ouroboros-network/pull/2526/commits/1f90e16c2a1d83c95f10f9654cda59c969cc5c84).  I think it makes sense to use `decrementBackoff` nad `incrementBackoff`.